eip-6170.md

```diff
@@ -0,0 +1,107 @@
+---
+eip: 6170
+title: Cross-Chain Messaging Interface
+description: A common smart contract interface for interacting with messaging protocols.
+author: Sujith Somraaj (@sujithsomraaj)
+discussions-to: https://ethereum-magicians.org/t/cross-chain-messaging-standard/12197
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-12-19
+---
+
+## Abstract
+
+This EIP standardizes an interface for cross-chain messengers, providing basic functionality to send and receive a cross-chain message (state).
+
+## Motivation
+
+Cross-chain messaging protocols lack standardization, resulting in unnecessarily complex competing implementations: Layerzero, Hyperlane & Wormhole each use a different interface. This makes integration difficult at the aggregator or plugin layer for protocols that must conform to any standards and forces each protocol to implement its adapter.
+
+Even chain-native arbitrary messaging protocols like the MATIC State Tunnel has an application-specific interface.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+Every compliant messaging tunnels must implement the following interface.
+
+``` solidity
+pragma solidity >=0.8.0;
+
+/// @title Cross-Chain Messaging interface
+/// @dev Allows seamless interchain messaging.
+/// @author Sujith Somraaj
+/// Note: Bytes are used along the entire implementation to support non-evm chains.
+
+interface EIP6170 {
+    /// @dev This emits when a message is sent from a sender on source chain to a receiver on a different blockchain.
+    /// @param _sender is the address of the sender.
+    /// @param _fromChainId is the identifier of the sending blockchain.
+    /// @param _receiver is the address of the receiver.
+    /// @param _toChainId is the identifier of the receiving blockchain.
+    /// @param _message is the arbitray message being sent.
+    /// NOTE: this emits when the message is sent, it does not gurantee successful delivery.
+    event MessageSent(
```

Comment:
These parameters will often want `index` for filtering, and this tends to be significant for off-chain functionality. So much so that folks will [emit two events](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/b450181c7fc255149ce86551f21dacd7ab47d5f5/solidity/contracts/Mailbox.sol#L84) just to get around the 3 index cap.